### ISSUE TITLE
[BUG] Whoogle Link redirecting to one with cookies disabled

### ISSUE THREAD

**Issue by Fxzzi** [benbusby/whoogle-search#749]
Hello, I am using a local instance of Whoogle. Navigating the to the page [http://localhost:5000/](http://localhost:5000/) Works completely fine, but if I try accessing it from my startpage: [https://fxzzi.github.io/startpage/](https://fxzzi.github.io/startpage/) It sends me to the link [http://localhost:5000/?cookies_disabled=1](http://localhost:5000/?cookies_disabled=1)

This occurs on Firefox and Librewolf. I am not sure whether it's a problem with Whoogle or Librewolf.

**Comment by hj-collab** [comment]
It occurs on all chromium browsers as well. This bug made Whoogle useless for me. Self hosting instance and configuring always rate limit my instance. Public instance works fine but I need to load my configuration every time. This is a MAJOR bug. I also tried modifying the search strings but to no avail.

**Comment by Fxzzi** [comment]
> This bug made Whoogle useless for me. 

This right here.

**Comment by onedr0p** [comment]
I think @benbusby should document what causes this to happen. What are the causes of cookies being disabled when you have this behind a reverse proxy. It's not very user friendly.

**Comment by navhawk80** [comment]
I found this bug since 0.7 version. This bug is not presented in old 0.6.0. version. So I'm using 0.6.0 till it will be fixed in 0.7.

**Comment by benbusby** [comment]
I just pushed a commit that should fix the issue documented here (accessing a Whoogle instance from an external link causing configuration/cookies to be disabled).

The cause of this was the `SameSite` attribute being set to `strict`, which disallows session persistence for requests initiated by a third party (see: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite#strict). This was originally implemented in a (perhaps misguided) attempt to improve overall security, but came at the cost of breaking functionality when interacting with the app from a third party context. This has been updated to use `lax` for the `SameSite` attribute, which will allow access from external links without disrupting configuration/sessions.

As far as how this `cookies_disabled` state is reached, session validation occurs as follows:
1. User visits Whoogle instance for the first time
2. Flask creates a session id for the user (ex/ "asdf123")
3. Flask initiates a temporary redirect to a `/session` endpoint, with the previous session id included as a URL param (i.e. `/session/asdf123`)
4. The Flask session is validated against the session ID provided in the URL
    - If the session from the URL param **does not** match the session that Flask created for that user, then the command to create the cookie for the user must have failed for some reason (typically this is seen when a user's environment is configured to block cookie creation). The `cookies_disabled` param is then appended to subsequent requests to avoid a situation where Flask continuously attempts creating a new session every time that user hits a Whoogle endpoint.
    - If the session from the URL param **does** match the session that Flask created for the user, then the cookie has successfully been set and the user can reliably have a unique configuration for that instance.
5. User is redirected to the appropriate page, with or without the `cookies_disabled` param attached

In short, this solution was devised as a way to create a mostly equal "session experience" for users and instance maintainers without A) relying on users to enable Javascript, B) causing runaway session creation for public instance maintainers, and C) allowing users with cookies blocked to still be able to use Whoogle, albeit in a diminished capacity. 

Let me know if anyone needs any clarification regarding anything I've mentioned.

**Comment by DUOLabs333** [comment]
Thanks! The commit referenced fixed an unrelated bug where the CSS sometimes didn't load. Not sure what cookies had to do with it though.

**Comment by hj-collab** [comment]
@benbusby Facing the same issue with 0.7.3. Any solution?
